README.md
# ServerspecExtendedTypes

[![Build Status](https://travis-ci.org/jantman/serverspec-extended-types.svg?branch=master)](https://travis-ci.org/jantman/serverspec-extended-types)
[![Code Coverage](https://codecov.io/github/jantman/serverspec-extended-types/coverage.svg?branch=master)](https://codecov.io/github/jantman/serverspec-extended-types?branch=master)
[![Code Climate](https://codeclimate.com/github/jantman/serverspec-extended-types/badges/gpa.svg)](https://codeclimate.com/github/jantman/serverspec-extended-types)
[![Gem Version](https://img.shields.io/gem/v/serverspec-extended-types.svg)](https://rubygems.org/gems/serverspec-extended-types)
[![Total Downloads](https://img.shields.io/gem/dt/serverspec-extended-types.svg)](https://rubygems.org/gems/serverspec-extended-types)
[![Github Issues](https://img.shields.io/github/issues/jantman/serverspec-extended-types.svg)](https://github.com/jantman/serverspec-extended-types/issues)
[![Project Status: Inactive – The project has reached a stable, usable state but is no longer being actively developed; support/maintenance will be provided as time allows.](http://www.repostatus.org/badges/2.0.0/inactive.svg)](http://www.repostatus.org/#inactive)

serverspec-extended-types provides some purpose-specific types to be used with [serverspec](http://serverspec.org/) 2.x for
testing various things on a host, as well as some types for high-level integration tests that make actual requests against
services on the host.

Current types include:

* A ``virtualenv`` type to make expectations about a Python Virtualenv, its ``pip`` and ``python`` versions, and packages
  installed in it. This also works for system-wide Python if ``pip`` is installed.
* A ``http_get`` type to perform an HTTP GET request (specifying a port number, the ``Host:`` header to set, and the path
  to request) and make expectations about whether the request times out, its status code, headers, response body, and
  JSON content (if the response body is parsable as JSON).
* A ``bitlbee`` type to actually connect via IRC to a running [bitlbee](http://www.bitlbee.org/) IRC gateway and authenticate
  to it, and make expectations that the connection and authentication worked, and what version of Bitlbee is running.

This is in no way associated with or endorsed by the SeverSpec project or its developers. (When I proposed that they include
a HTTP request type, I was told that was "using [serverspec] wrong", and the GitHub issue was deleted). I imagine that,
at the least, they'd tell me that (1) these should be in SpecInfra, and (2) these aren't "proper" things for ServerSpec
to test. That being said, these are useful to me, for my purposes. I hope they're also useful to someone else.

## Installation

Add this line to your application's Gemfile:

    gem 'serverspec-extended-types'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install serverspec-extended-types

### Requirements

* [serverspec](https://rubygems.org/gems/serverspec) 2.x
* [specinfra](https://rubygems.org/gems/specinfra) 2.x

## Usage

In your spec_helper, add a line like:

    require 'serverspec_extended_types'

Then use the various types that this gem provides:

## Types

### bitlbee

The bitlbee type allows testing connectivity to a [bitlbee](http://www.bitlbee.org/) IRC gateway server.

    describe bitlbee(port, nick, password, use_ssl=false) do
      # matchers here
	end

All server communication happens during instantiation of the type; when ``describe bitlbee()`` is executed,
the type will attempt to connect to the server (with a timeout of 10 seconds) and authenticate as
the specified user, and store a number of state variables for later use by the various matchers.

This type supports SSL, but connects to the server without verifying certificates.

#### Parameters:

* __port__ - integer port number to connect to
* __nick__ - nick/username to login as
* __password__ - the password for nick
* __use_ssl__ - Boolean, whether or not to use SSL; defaults to false

#### Matchers

##### connectable?

True if the test was able to successfully connect and authenticate, false otherwise.

    describe bitlbee(port, nick, password) do
      it { should be_connectable }
	end

##### timed_out?

True if the connection is aborted by a timeout, false otherwise.

    describe bitlbee(port, nick, password) do
      it { should_not be_timed_out }
	end

##### version

Returns the Bitlbee version String.

    describe bitlbee(port, nick, password) do
      its(:version) { should match /3\.\d+\.\d+ Linux/ }
	end

### http_get

The http_get type performs an HTTP GET from the local (rspec runner) system, against
the IP address that serverspec is running against (``ENV[TARGET_HOST]``), with a
specified ``Host`` header value. The request is wrapped in a configurable-length timeout.

    describe http_get(port, host_header, path, timeout_sec=10, protocol='http', bypass_ssl_verify=false)
      # matchers here
	end

All server communication happens during instantiation of the type; when ``describe http_get()`` is executed,
the type will attempt to issue the GET request with a default timeout of 10 seconds, and store a number of
state variables for later use by the various matchers.

#### Parameters

* __port__ - the port to make the HTTP request to
* __host_header__ - the ``Host`` header value to provide in the request
* __path__ - the path to request from the server
* __timeout_sec__ - timeout in seconds before canceling the request (Int; default 10)
* __protocol__ - protocol to be used (default to `http`,  can be `http`)
* __bypass_ssl_verify__ - bypass SSL verification (default to `false` to keep good security. Set it to `true` for self-signed certificates only!)

#### Matchers

##### body

Returns the String body content of the HTTP response.

    describe http_get(80, 'myhostname', '/') do
      its(:body) { should match /<html>/ }
    end

##### headers

Returns the HTTP response headers as a hash.

    describe http_get(80, 'myhostname', '/') do
      its(:headers) { should include('HeaderName' => /value regex/) }
    end

##### json

If the response body is deserializable as JSON, contains the JSON hash,
otherwise an empty hash.

    describe http_get(80, 'myhostname', '/') do
      its(:json) { should include('Key' => /value regex/) }
    end

##### status

Returns the HTTP status code, or 0 if timed out.

    describe http_get(80, 'myhostname', '/') do
      its(:status) { should eq 200 }
    end

##### timed_out?

True if the request timed out, false otherwise.

    describe http_get(80, 'myhostname', '/') do
      it { should_not be_timed_out }
    end

##### redirected?

True if the request was redirected, false otherwise.

    describe http_get(80, 'myhostname', '/') do
      it { should be_redirected }
    end

##### redirected_to?

True if the request was redirected to the specified location, false otherwise.

    describe http_get(80, 'myhostname', '/') do
      it { should be_redirected_to 'https://myhostname/newpath' }
    end

    describe http_get(80, 'myhostname', '/') do
      it { should be_redirected_to 'http://mynewhost/' }
    end

### virtualenv

The virtualenv type has various matchers for testing the state and content of a
Python [virtualenv](https://virtualenv.pypa.io/en/latest/). It executes commands
via the builtin serverspec/specinfra command execution (i.e. uses the same backend
code that the built-in ``command`` and ``file`` types use).

    describe virtualenv('/path/to/venv') do
      # matchers here
    end

Unlike the ``http_get`` and ``bitlbee`` types, execution of the ``virtualenv`` type
commands is triggered by the matchers rather than the ``describe`` clause.

#### Parameters

* __name__ - the absolute path to the root of the virtualenv on the filesystem

#### Matchers

##### pip_freeze

Return a hash of all packages present in `pip freeze` output for the venv

Note that any editable packages (`-e something`) are returned as hash keys
with an empty value.

    describe virtualenv('/path/to/venv') do
      its(:pip_freeze) { should include('wsgiref' => '0.1.2') }
      its(:pip_freeze) { should include('requests') }
      its(:pip_freeze) { should include('pytest' => /^2\.6/) }
      its(:pip_freeze) { should include('-e git+git@github.com:jantman/someproject.git@1d8a380e3af9d081081d7ef685979200a7db4130#egg=someproject') }
    end

##### pip_version

Return the version of pip installed in the virtualenv

    describe virtualenv('/path/to/venv') do
      its(:pip_version) { should match /^6\.0\.6$/ }
    end

##### python_version

Return the version of python installed in the virtualenv

    describe virtualenv('/path/to/venv') do
      its(:python_version) { should match /^2\.7\.9$/ }
    end

##### virtualenv?

Test whether this appears to be a working venv

    describe virtualenv('/path/to/venv') do
      it { should be_virtualenv }
    end

Tests performed:

* venv_path/bin/pip executable by owner?
* venv_path/bin/python executable by owner?
* venv_path/bin/activate readable by owner?
* 'export VIRTUAL_ENV' in venv_path/bin/activate?

## Contributing

1. Fork it ( https://github.com/jantman/serverspec-extended-types/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request

## Testing

Spec tests are done automatically via Travis CI. They're run using Bundler and rspec.

For manual testing:

    bundle install
    bundle exec rake test

## Releasing

1. Ensure all tests are passing, coverage is acceptable, etc.
2. Increment the version number in ``lib/serverspec_extended_types/version.rb``
3. Update CHANGES.md
4. Push those changes to origin.
5. ``bundle exec rake build``
6. ``bundle exec rake release``


CHANGES.md
# serverspec-extended-types, changelog

## 0.1.1 2017-07-28 ojizero <z.acnts.816@protonmail.ch>

* Add `redirected_to` and `redirected` matchers for http_get type.

## 0.1.0 2017-05-15 Jason Antman <jason@jasonantman.com>

* Add `protocol` (http|https) and `bypass_ssl_verify` parameters for http_get.

## 0.0.3 2015-03-23 Jason Antman <jason@jasonantman.com>

* Add ``json`` matcher for http_get type.

## 0.0.2 2015-03-14 Jason Antman <jason@jasonantman.com>

* Minor bugfix in Serverspec::Types::Virtualenv#virtualenv?

## 0.0.1 2015-03-14 Jason Antman <jason@jasonantman.com>

* Initial release with bitlbee, http_get and virtualenv types


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path('../../lib', __FILE__)

require 'simplecov'
SimpleCov.start do
  add_filter "/vendor/"
end

require 'codecov'
if ENV['CI']=='true'
  SimpleCov.formatter = SimpleCov::Formatter::Codecov
end

require 'specinfra'
require 'rspec/mocks/standalone'
require 'rspec/its'
require 'specinfra/helper/set'
include Specinfra::Helper::Set

set :backend, :exec

set :os, :family => 'linux'

module Specinfra
  module Backend
    class Ssh
      def run_command(cmd, opts={})
        CommandResult.new :stdout => nil, :exit_status => 0
      end
    end
  end
end

module GetCommand
  def get_command(method, *args)
    Specinfra.command.get(method, *args)
  end
end

include GetCommand


require 'serverspec_extended_types'


spec/types/bitlbee_spec.rb
require 'spec_helper'
require 'serverspec_extended_types/http_get'

describe 'Serverspec::Type.bitlbee' do
  it 'instantiates the class with the correct parameters' do
    expect(Serverspec::Type::Bitlbee).to receive(:new).with(1, 'mynick', 'mypass', false)
    bitlbee(1, 'mynick', 'mypass')
  end
  it 'returns the new object' do
    expect(Serverspec::Type::Bitlbee).to receive(:new).and_return("foo")
    expect(bitlbee(1, 'foo', 'bar')).to eq "foo"
  end
end
describe 'Serverspec::Type::Bitlbee' do
  context '#initialize' do
    it 'sets instance variables' do
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))
      x = bitlbee(1, 'foo', 'bar')
      expect(x.instance_variable_get(:@port)).to eq 1
      expect(x.instance_variable_get(:@host)).to eq 'myhost'
      expect(x.instance_variable_get(:@nick)).to eq 'foo'
      expect(x.instance_variable_get(:@password)).to eq 'bar'
      expect(x.instance_variable_get(:@connected_status)).to eq false
      expect(x.instance_variable_get(:@version_str)).to eq ''
      expect(x.instance_variable_get(:@timed_out_status)).to eq false
      expect(x.instance_variable_get(:@started)).to eq false
      expect(x.instance_variable_get(:@use_ssl)).to eq false
    end
    it 'sets use_ssl instance variable' do
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))
      x = bitlbee(1, 'foo', 'bar', use_ssl=true)
      expect(x.instance_variable_get(:@use_ssl)).to eq true
    end
  end
  context 'start' do
    it 'calls connect_ssl if use_ssl=true' do
      x = bitlbee(1, 'foo', 'bar', use_ssl=true)
      expect(x).to receive(:connect_ssl).once
      expect(x).to_not receive(:connect)
      x.start
      expect(x.instance_variable_get(:@timed_out_status)).to eq false
      expect(x.instance_variable_get(:@connected_status)).to eq false
      expect(x.instance_variable_get(:@started)).to eq true
    end
    it 'calls connect if use_ssl=false' do
      x = bitlbee(1, 'foo', 'bar')
      expect(x).to receive(:connect).once
      expect(x).to_not receive(:connect_ssl)
      x.start
      expect(x.instance_variable_get(:@timed_out_status)).to eq false
      expect(x.instance_variable_get(:@connected_status)).to eq false
      expect(x.instance_variable_get(:@started)).to eq true
    end
    it 'runs with a timeout' do
      x = bitlbee(1, 'foo', 'bar')
      expect(Timeout).to receive(:timeout).with(10)
      x.start
      expect(x.instance_variable_get(:@timed_out_status)).to eq false
      expect(x.instance_variable_get(:@connected_status)).to eq false
      expect(x.instance_variable_get(:@started)).to eq true
      expect(x.timed_out?).to eq false
    end
    it 'sets timed_out_status on timeout' do
      x = bitlbee(1, 'foo', 'bar')
      expect(Timeout).to receive(:timeout).with(10).and_raise(Timeout::Error)
      x.start
      expect(x.instance_variable_get(:@timed_out_status)).to eq true
      expect(x.instance_variable_get(:@connected_status)).to eq false
      expect(x.instance_variable_get(:@started)).to eq true
      expect(x.timed_out?).to eq true
      expect(x.connectable?).to eq false
    end
    it 'sets connected_status on ECONNREFUSED' do
      x = bitlbee(1, 'foo', 'bar')
      expect(x).to receive(:connect).and_raise(Errno::ECONNREFUSED)
      x.start
      expect(x.instance_variable_get(:@timed_out_status)).to eq false
      expect(x.instance_variable_get(:@connected_status)).to eq false
      expect(x.instance_variable_get(:@started)).to eq true
      expect(x.timed_out?).to eq false
      expect(x.connectable?).to eq false
    end
  end
  context '#connect_ssl' do
    it 'calls communicate' do
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))
      sock_dbl = double
      expect(TCPSocket).to receive(:open).with('myhost', 1).and_return(sock_dbl)
      sslctx_dbl = double
      expect(OpenSSL::SSL::SSLContext).to receive(:new).and_return(sslctx_dbl)
      expect(sslctx_dbl).to receive(:set_params).with(verify_mode: OpenSSL::SSL::VERIFY_NONE).ordered
      sslsock_dbl = double
      expect(sslsock_dbl).to receive(:sync_close=).ordered.with(true)
      expect(sslsock_dbl).to receive(:connect).ordered
      expect(sslsock_dbl).to receive(:puts).ordered.with("communicate")
      expect(sslsock_dbl).to receive(:puts).ordered.with("QUIT :\"outta here\"\n")
      expect(sslsock_dbl).to receive(:close).ordered
      expect(OpenSSL::SSL::SSLSocket).to receive(:new).with(sock_dbl, sslctx_dbl).and_return(sslsock_dbl)
      # expect_any_instance_of(Serverspec::Type::Bitlbee).to receive(:communicate)
      x = bitlbee(1, 'foo', 'bar')
      allow(x).to receive(:communicate) do
        x.instance_variable_get(:@socket).puts("communicate")
      end
      x.connect_ssl
      expect(x.instance_variable_get(:@socket)).to eq sslsock_dbl
    end
  end
  context '#connect' do
    it 'calls communicate' do
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))
      sock_dbl = double
      expect(TCPSocket).to receive(:open).with('myhost', 1).and_return(sock_dbl)
      expect(sock_dbl).to receive(:puts).ordered.with("communicate")
      expect(sock_dbl).to receive(:puts).ordered.with("QUIT :\"outta here\"\n")
      expect(sock_dbl).to receive(:close).ordered
      x = bitlbee(1, 'foo', 'bar')
      allow(x).to receive(:communicate) do
        x.instance_variable_get(:@socket).puts("communicate")
      end
      x.connect
      expect(x.instance_variable_get(:@socket)).to eq sock_dbl
    end
  end
  context '#communicate' do
    it 'logs in' do
      x = bitlbee(1, 'foo', 'bar')
      sock = double
      x.instance_variable_set(:@socket, sock)
      x.instance_variable_set(:@started, true)
      expect(sock).to receive(:puts).ordered.with("PASS bar\n")
      expect(sock).to receive(:puts).ordered.with("NICK foo\n")
      expect(sock).to receive(:puts).ordered.with("USER foo foo servername :TestUser\n")
      expect(sock).to receive(:readpartial).ordered.with(1024).and_return("Welcome to the foo")
      expect(sock).to receive(:puts).ordered.with("MODE foo +i\n")
      expect(sock).to receive(:readpartial).ordered.with(1024).and_return("If you've never something something")
      expect(sock).to receive(:puts).ordered.with("PRIVMSG &bitlbee :identify bar\n")
      expect(sock).to receive(:readpartial).ordered.with(1024).and_return("PING servername")
      expect(sock).to receive(:puts).ordered.with(":servername PONG servername :servername\n")
      expect(sock).to receive(:readpartial).ordered.with(1024).and_return("MODE foo :\+i")
      expect(sock).to receive(:puts).ordered.with("PRIVMSG root :\001VERSION\001\n")
      expect(sock).to receive(:readpartial).ordered.with(1024).and_return("foo bar baz blam\n")
      expect(sock).to receive(:readpartial).ordered.with(1024).and_return("VERSION x.y.z\n")
      x.communicate
      expect(x.instance_variable_get(:@connected_status)).to eq true
      expect(x.version).to eq 'x.y.z'
      expect(x.connectable?).to eq true
      expect(x.timed_out?).to eq false
    end
  end
  context '#timed_out?' do
    it 'calls start if not started' do
      x = bitlbee(1, 'foo', 'bar')
      expect(x).to receive(:start).once
      expect(x.timed_out?).to eq false
    end
    it 'doesnt call start if started' do
      x = bitlbee(1, 'foo', 'bar')
      x.instance_variable_set(:@started, true)
      expect(x).to_not receive(:start)
      expect(x.timed_out?).to eq false
    end
  end
  context '#connectable?' do
    it 'calls start if not started' do
      x = bitlbee(1, 'foo', 'bar')
      expect(x).to receive(:start).once
      expect(x.connectable?).to eq false
    end
    it 'doesnt call start if started' do
      x = bitlbee(1, 'foo', 'bar')
      x.instance_variable_set(:@started, true)
      expect(x).to_not receive(:start)
      expect(x.connectable?).to eq false
    end
  end
  context '#version?' do
    it 'calls start if not started' do
      x = bitlbee(1, 'foo', 'bar')
      expect(x).to receive(:start).once
      expect(x.version).to eq ""
    end
    it 'doesnt call start if started' do
      x = bitlbee(1, 'foo', 'bar')
      x.instance_variable_set(:@started, true)
      expect(x).to_not receive(:start)
      expect(x.version).to eq ""
    end
  end
end


spec/types/http_get_spec.rb
require 'spec_helper'
require 'serverspec_extended_types/http_get'

describe 'http_get()' do
  context 'Serverspec::Type method' do
    it 'instantiates the class with the correct parameters' do
      expect(Serverspec::Type::Http_Get).to receive(:new).with(1, 'hostheader', 'mypath', timeout_sec=20, 'http', false)
      http_get(1, 'hostheader', 'mypath', timeout_sec=20)
    end
    it 'returns the new object' do
      expect(Serverspec::Type::Http_Get).to receive(:new).and_return("foo")
      expect(http_get(1, 'hostheader', 'mypath', timeout_sec=20)).to eq "foo"
    end
  end
  context 'initialize' do
    it 'sets instance variables' do
      expect_any_instance_of(Serverspec::Type::Http_Get).to receive(:getpage).once
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))
      x = http_get(1, 'hostheader', 'mypath', timeout_sec=20)
      expect(x.instance_variable_get(:@ip)).to eq 'myhost'
      expect(x.instance_variable_get(:@port)).to eq 1
      expect(x.instance_variable_get(:@host)).to eq 'hostheader'
      expect(x.instance_variable_get(:@path)).to eq 'mypath'
      expect(x.instance_variable_get(:@timed_out_status)).to eq false
      expect(x.instance_variable_get(:@content_str)).to be nil
      expect(x.instance_variable_get(:@headers_hash)).to be nil
      expect(x.instance_variable_get(:@response_code_int)).to be nil
      expect(x.instance_variable_get(:@response_json)).to be nil
      expect(x.instance_variable_get(:@protocol)).to eq 'http'
      expect(x.instance_variable_get(:@bypass_ssl_verify)).to eq false
      expect(x.instance_variable_get(:@redirects)).to eq false
      expect(x.instance_variable_get(:@redirect_path)).to be nil
    end
    it 'calls getpage' do
      expect_any_instance_of(Serverspec::Type::Http_Get).to receive(:getpage).once
      x = http_get(1, 'hostheader', 'mypath', timeout_sec=20)
      expect(x.instance_variable_get(:@timed_out_status)).to eq false
    end
    it 'runs with a timeout' do
      expect(Timeout).to receive(:timeout).with(10)
      x = http_get(1, 'hostheader', 'mypath')
      expect(x.instance_variable_get(:@timed_out_status)).to eq false
    end
    it 'runs with a timeout with the specified length' do
      expect(Timeout).to receive(:timeout).with(20)
      x = http_get(1, 'hostheader', 'mypath', timeout_sec=20)
      expect(x.instance_variable_get(:@timed_out_status)).to eq false
    end
    it 'sets timed_out_status on timeout' do
      expect(Timeout).to receive(:timeout).with(20).and_raise(Timeout::Error)
      x = http_get(1, 'hostheader', 'mypath', timeout_sec=20)
      expect(x.instance_variable_get(:@timed_out_status)).to eq true
      expect(x.timed_out?).to eq true
      expect(x.status).to eq 0
    end
  end
  context '#getpage' do
    it 'requests the page' do
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))
      conn = double
      headers = double
      expect(headers).to receive(:[]=).with(:user_agent, %r"Serverspec::Type::Http_Get/\d+\.\d+\.\d+ \(https://github.com/jantman/serverspec-extended-types\)")
      expect(conn).to receive(:headers).and_return(headers)
      expect(headers).to receive(:[]=).with(:Host, 'hostheader')
      expect(conn).to receive(:headers).and_return(headers)
      response = double
      expect(response).to receive(:status).and_return(200)
      expect(response).to receive(:body).and_return("foo bar")
      expect(response).to receive(:headers).and_return({'h1' => 'h1val', 'h2' => 'h2val'})
      expect(conn).to receive(:get).with('mypath').and_return(response)
      expect(Faraday).to receive(:new).with("http://myhost:1/").and_return(conn)
      x = http_get(1, 'hostheader', 'mypath')
      expect(x.timed_out?).to eq false
      expect(x.status).to eq 200
      expect(x.body).to eq 'foo bar'
      expected_headers = {'h1' => 'h1val', 'h2' => 'h2val'}
      expect(x.headers).to eq expected_headers
      expect(x.json).to be_empty
      expect(x.json).to be_a_kind_of(Hash)
    end
    it 'supports https' do
      # boilerplate
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))
      headers = double
      expect(headers).to receive(:[]=).with(:user_agent, %r"Serverspec::Type::Http_Get/\d+\.\d+\.\d+ \(https://github.com/jantman/serverspec-extended-types\)")
      expect(headers).to receive(:[]=).with(:Host, 'hostheader')
      conn = double(headers: headers)
      response = double(status: 200, body: "OK", headers: [])
      expect(conn).to receive(:get).with('mypath').and_return(response)
      # most importantly, we want https here
      expect(Faraday).to receive(:new).with("https://myhost:1/").and_return(conn)
      x = http_get(1, 'hostheader', 'mypath', 30, 'https')
    end
    it 'supports ssl verify bypass for self-signed certificates' do
      # boilerplate
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))
      headers = double
      expect(headers).to receive(:[]=).with(:user_agent, %r"Serverspec::Type::Http_Get/\d+\.\d+\.\d+ \(https://github.com/jantman/serverspec-extended-types\)")
      expect(headers).to receive(:[]=).with(:Host, 'hostheader')
      conn = double(headers: headers)
      response = double(status: 200, body: "OK", headers: [])
      expect(conn).to receive(:get).with('mypath').and_return(response)
      # most importantly, we want https here
      expect(Faraday).to receive(:new).with("https://myhost:1/", {ssl: {verify: false}}).and_return(conn)
      x = http_get(1, 'hostheader', 'mypath', 30, 'https', true)
    end
    it 'sets JSON if parsable' do
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))
      conn = double
      headers = double
      expect(headers).to receive(:[]=).with(:user_agent, %r"Serverspec::Type::Http_Get/\d+\.\d+\.\d+ \(https://github.com/jantman/serverspec-extended-types\)")
      expect(conn).to receive(:headers).and_return(headers)
      expect(headers).to receive(:[]=).with(:Host, 'hostheader')
      expect(conn).to receive(:headers).and_return(headers)
      response = double
      expect(response).to receive(:status).and_return(200)
      expect(response).to receive(:body).and_return('{"foo": "bar", "baz": {"blam": "blarg"}}')
      expect(response).to receive(:headers).and_return({'h1' => 'h1val', 'h2' => 'h2val'})
      expect(conn).to receive(:get).with('mypath').and_return(response)
      expect(Faraday).to receive(:new).with("http://myhost:1/").and_return(conn)
      x = http_get(1, 'hostheader', 'mypath')
      expect(x.timed_out?).to eq false
      expect(x.status).to eq 200
      expect(x.body).to eq '{"foo": "bar", "baz": {"blam": "blarg"}}'
      expected_headers = {'h1' => 'h1val', 'h2' => 'h2val'}
      expect(x.headers).to eq expected_headers
      expected_json = {"foo" => "bar", "baz" => {"blam" => "blarg"}}
      expect(x.json).to eq expected_json
    end
    it 'without redirects' do
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))

      headers = double('header')
      conn = double('conn', headers: headers)

      expect(headers).to receive(:[]=).with(:user_agent, %r"Serverspec::Type::Http_Get/\d+\.\d+\.\d+ \(https://github.com/jantman/serverspec-extended-types\)")
      expect(headers).to receive(:[]=).with(:Host, 'hostheader')

      response = double('resp', status: 200, body: 'OK', headers: ({}))

      expect(conn).to receive(:get).with('mypath').and_return(response)

      expect(Faraday).to receive(:new).with('http://myhost:1/').and_return(conn)
      x = http_get(1, 'hostheader', 'mypath')

      # test internal parameters
      expect(x.instance_variable_get(:@redirects)).to eq false
      expect(x.instance_variable_get(:@redirect_path)).to be nil

      # test exposed API
      expect(x).to_not be_redirected
      expect(x).to_not be_redirected_to 'https://myhost:1/mynewpath'
    end
    it 'supports redirects' do
      stub_const('ENV', ENV.to_hash.merge('TARGET_HOST' => 'myhost'))

      headers = double('header')
      conn = double('conn', headers: headers)

      expect(headers).to receive(:[]=).with(:user_agent, %r"Serverspec::Type::Http_Get/\d+\.\d+\.\d+ \(https://github.com/jantman/serverspec-extended-types\)")
      expect(headers).to receive(:[]=).with(:Host, 'hostheader')

      redirect_location = 'https://myhost:1/mynewpath'
      response = double('resp', status: 301, body: 'OK', headers: {'location' => redirect_location})

      expect(conn).to receive(:get).with('mypath').and_return(response)

      expect(Faraday).to receive(:new).with('http://myhost:1/').and_return(conn)
      x = http_get(1, 'hostheader', 'mypath')

      # test internal parameters
      expect(x.instance_variable_get(:@redirects)).to eq true
      expect(x.instance_variable_get(:@redirect_path)).to_not be nil
      expect(x.instance_variable_get(:@redirect_path)).to eq redirect_location

      # test exposed API
      expect(x).to be_redirected
      expect(x).to be_redirected_to redirect_location
    end
  end
end


lib/serverspec_extended_types/http_get.rb
##############################################################################
# serverspec-extended-types - http_get
#
# <https://github.com/jantman/serverspec-extended-types>
#
# Copyright (C) 2015 Jason Antman <jason@jasonantman.com>
#
# Licensed under the MIT License - see LICENSE.txt
#
##############################################################################

require 'timeout'
require 'faraday'
require 'serverspec_extended_types/version'
require 'json'

module Serverspec
  module Type

    # Perform an HTTP GET request against the serverspec target
    # using {http://www.rubydoc.info/gems/faraday/ Faraday}.
    class Http_Get < Base
      # Http status codes that are tested against for redirect test
      @@redirect_codes = [301, 302, 307, 308]

      # Initialize a bunch of instance variables, then call {#getpage}
      #
      # Calls {#getpage} within a {Timeout::timeout} block;
      # If {#getpage} times out, set ``timed_out_status`` to [True]
      #
      # @param port [Int] the port to connect to HTTP over
      # @param host_header [String] the value to set in the 'Host' HTTP request header
      # @param path [String] the URI/path to request from the server
      # @param timeout_sec [Int] how many seconds to allow request to run before timing out and setting @timed_out_status to True
      # @param protocol [String] the protocol to connect to the server (e.g. 'http', 'https')
      # @param bypass_ssl_verify [Boolean] if true, SSL verification will be bypassed (useful for self-signed certificates)
      #
      # @example
      #   describe http_get(80, 'myhostname', '/') do
      #     # tests here
      #   end
      #
      # @api public
      # @return [nil]
      def initialize(port, host_header, path, timeout_sec=10, protocol='http', bypass_ssl_verify=false)
        @ip = ENV['TARGET_HOST']
        @port = port
        @host = host_header
        @path = path
        @timed_out_status = false
        @content_str = nil
        @headers_hash = nil
        @response_code_int = nil
        @response_json = nil
        @protocol = protocol
        @bypass_ssl_verify = bypass_ssl_verify
        @redirects = false
        @redirect_path = nil
        begin
          Timeout::timeout(timeout_sec) do
            getpage
          end
        rescue Timeout::Error
          @timed_out_status = true
        end
      end

      # Private method to actually get the page; must be called within a timeout block
      #
      # Gets the page using {http://www.rubydoc.info/gems/faraday/ Faraday} and then sets instance variables for the various attribute readers.
      #
      # @api private
      # @return [nil]
      def getpage
        ip = @ip
        port = @port
        protocol = @protocol
        options = []
        options << { ssl: { verify: false } } if @bypass_ssl_verify
        conn = Faraday.new("#{protocol}://#{ip}:#{port}/", *options)
        version = ServerspecExtendedTypes::VERSION
        conn.headers[:user_agent] = "Serverspec::Type::Http_Get/#{version} (https://github.com/jantman/serverspec-extended-types)"
        conn.headers[:Host] = @host
        response = conn.get(@path)
        @response_code_int = response.status
        @content_str = response.body
        @headers_hash = Hash.new('')
        response.headers.each do |header, val|
          @headers_hash[header] = val
        end
        @redirects = @@redirect_codes.include? @response_code_int
        @redirect_path = @redirects ? @headers_hash['location'] : nil
        # try to JSON decode
        begin
          @response_json = JSON.parse(@content_str)
        rescue
          @response_json = {}
        end
      end

      # Whether or not the request timed out
      #
      # @example
      #   describe http_get(80, 'myhostname', '/') do
      #     it { should_not be_timed_out }
      #   end
      #
      # @api public
      # @return [Boolean]
      def timed_out?
        @timed_out_status
      end

      # Returns the HTTP response headers as a hash
      #
      # @example
      #   describe http_get(80, 'myhostname', '/') do
      #     its(:headers) { should include('HeaderName' => /value regex/) }
      #   end
      #
      # @api public
      # @return [Hash]
      def headers
        @headers_hash
      end

      # Returns the decoded JSON response, or an empty hash
      #
      # @example
      #   describe http_get(80, 'myhostname', '/') do
      #     its(:json) { should include('foo' => /value regex/) }
      #   end
      #
      # @api public
      # @return [Hash]
      def json
        @response_json
      end

      # Returns the HTTP status code, or 0 if timed out
      #
      # @example
      #   describe http_get(80, 'myhostname', '/') do
      #     its(:status) { should eq 200 }
      #   end
      #
      # @api public
      # @return [Int]
      def status
        if @timed_out_status
          0
        else
          @response_code_int
        end
      end

      # Returns the body/content of the HTTP response
      #
      # @example
      #   describe http_get(80, 'myhostname', '/') do
      #     its(:body) { should match /<html>/ }
      #   end
      #
      # @api public
      # @return [String]
      def body
        @content_str
      end

      # Whether or not it redirects to some other page
      #
      # @example
      #   describe http_get(80, 'myhostname', '/') do
      #     it { should be_redirected_to 'https://myhostname/' }
      #   end
      #
      # @api public
      # @return [Boolean]
      def redirected_to? (redirect_path)
        @redirects and @redirect_path == redirect_path
      end

      # Whether or not it redirects to any other page
      #
      # @example
      #   describe http_get(80, 'myhostname', '/') do
      #     it { should be_redirected }
      #   end
      #
      # @api public
      # @return [Boolean]
      def redirected?
        @redirects
      end

      private :getpage
    end

    # ServerSpec Type wrapper for http_get
    #
    # @example
    #   describe http_get(80, 'myhostname', '/') do
    #     # tests here
    #   end
    #
    # @param port [Int] the port to connect to HTTP over
    # @param host_header [String] the value to set in the 'Host' HTTP request header
    # @param path [String] the URI/path to request from the server
    # @param timeout_sec [Int] how many seconds to allow request to run before timing out and setting @timed_out_status to True
    # @param protocol [String] the protocol to connect to the server (default 'http', can be 'https')
    # @param bypass_ssl_verify [Boolean] if true, SSL verification will be bypassed (useful for self-signed certificates)
    #
    # @api public
    # @return [Serverspec::Type::Http_Get]
    def http_get(port, host_header, path, timeout_sec=10, protocol='http', bypass_ssl_verify=false)
      Http_Get.new(port, host_header, path, timeout_sec, protocol, bypass_ssl_verify)
    end
  end
end

include Serverspec::Type
